servers/rendering/renderer_rd/effects/metal_fx.mm

```diff
@@ -0,0 +1,214 @@
+/**************************************************************************/
+/*  metal_fx.mm                                                           */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#import "metal_fx.h"
+
+#import "drivers/metal/pixel_formats.h"
+#import "drivers/metal/rendering_device_driver_metal.h"
+
+#import <Metal/Metal.h>
+#import <MetalFX/MetalFX.h>
+
+using namespace RendererRD;
+
+#pragma mark - Spatial Scaler
+
+MFXSpatialContext::~MFXSpatialContext() {
+}
+
+MFXSpatialEffect::MFXSpatialEffect() {
+}
+
+MFXSpatialEffect::~MFXSpatialEffect() {
+}
+
+void MFXSpatialEffect::callback(RDD *p_driver, RDD::CommandBufferID p_command_buffer, CallbackArgs *p_userdata) {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability"
+
+	MDCommandBuffer *obj = (MDCommandBuffer *)(p_command_buffer.id);
+	obj->end();
```

Comment:
Is this making some sort of assumption of the command buffer already being started? That sounds alright, but will the command buffer remain open or reopen itself after the scaler is finished for other commands that might come afterwards? My main concern is if it's flexible enough against the reorde that might occur.

Reply:
This is an implementation detail of Metal and the `MDCommandBuffer`, as command buffers can't create new encoders if one is already outstanding. I could change the `MDCommandBuffer` to expose a method called `execute_spatial_upscale` and then it would call `end` automatically as an internal detail, like it does when binding a pipeline:

https://github.com/godotengine/godot/blob/76128893e0c2b1d5db589bced648098831bb2d18/drivers/metal/metal_objects.mm#L87-L96

I could also create one for `execute_temporal_upscale`.

Reply:
I think it's good as long as it's aware that it needs to restart the buffer afterwards in later calls, I was just unsure and wanted to double check.